### PR TITLE
ADAPT-3235: Add PRO FontAwesome icon support

### PR DESCRIPTION
**Pull request by yvonnetangsu**
# READY FOR REVIEW

# Summary
- Add support for PRO icons

# Review By (Date)
- When does this need to be reviewed by?

# Criticality
- How critical is this PR on a 1-10 scale? Also see [Severity Assessment](https://stanfordits.atlassian.net/browse/D8CORE-1705).
- E.g., it affects one site, or every site and product?

# Review Tasks

## Setup tasks and/or behavior to test

1. Go to https://deploy-preview-148--stanford-alumni.netlify.app/test/yvonne/test-basic-page/
2. Check that the PRO icons are loading in the CTA links
Screen Shot 2021-09-09 at 6 37 13 PM
Screen Shot 2021-09-10 at 1 19 50 AM

3. https://deploy-preview-148--stanford-alumni.netlify.app/test/yvonne/test-dark-page-with-hero/
4. Check that the PRO icons are loading in the icon cards
Screen Shot 2021-09-09 at 6 40 10 PM

## Front End Validation
- [ ] Is the markup using the appropriate semantic tags and passes HTML validation?
- [ ] Cross-browser testing has been performed?
- [ ] Automated accessibility scans performed?
- [ ] Manual accessibility tests performed?
- [ ] Design is approved by @ user?

### Code
- [ ] Are the naming conventions following our standards?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided? eg (unit, behat, or codeception)

### Code security
- [ ] Are all [forms properly sanitized](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

## General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?

# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules?

# Associated Issues and/or People
- JIRA ticket(s)
- Other PRs
- Any other contextual information that might be helpful (e.g., description of a bug that this PR fixes, new functionality that it adds, etc.)
- Anyone who should be notified? (`@mention` them here)

# Resources
- [AMP Tool](https://stanford.levelaccess.net/index.php)
- [Accessibility Manual Test Script](https://docs.google.com/document/d/1ZXJ9RIUNXsS674ow9j3qJ2g1OAkCjmqMXl0Gs8XHEPQ/edit?usp=sharing)
- [HTML Validator](https://validator.w3.org/)
- [Browserstack](https://live.browserstack.com/dashboard) and link to [Browserstack Credentials](https://asconfluence.stanford.edu/confluence/display/SWS/External+Account+Credentials)

### CODE REVIEW

**Comment by yvonnetangsu** [comment]
Screen Shot 2021-08-31 at 4 41 59 PM

@sherakama , this builds but with some issues. I installed FontAwesome pro using the method here under "environment variables".
https://fontawesome.com/v5.15/how-to-use/on-the-web/setup/using-package-managers
I added the token in my env files. For the netlify side, do I need to go to netlify and replace the build script with this `npm config set '//registry.npmjs.org/:_authToken' "${GATSBY_FA_NPM_AUTH_TOKEN}" && npm run build`

https://stackoverflow.com/questions/52015748/npm-failed-to-replace-env-in-config-npm-token

**Comment by yvonnetangsu** [comment]
@mattanglin Tagging you as well - I know you're pretty busy, but in case you're already familiar with this issue 😄

**Comment by mattanglin** [comment]
@yvonnetangsu 

> Screen Shot 2021-08-31 at 4 41 59 PM
> 
> @sherakama , this builds but with some issues. I installed FontAwesome pro using the method here under "environment variables".
> 
> Screen Shot 2021-08-31 at 4 46 20 PM
> 
> https://fontawesome.com/v5.15/how-to-use/on-the-web/setup/using-package-managers
> 
> I added the token in my env files. For the netlify side, do I need to go to netlify and replace the build script with this `npm config set '//registry.npmjs.org/:_authToken' "${GATSBY_FA_NPM_AUTH_TOKEN}" && npm run build`? I have added the FA token env variable on Netlify but that alone still cause the error above.
> 
> https://stackoverflow.com/questions/52015748/npm-failed-to-replace-env-in-config-npm-token

@yvonnetangsu Were you able to resolve this npm install error/warning? I_ believe_ that since netlify takes care of package installation we need to [follow their naming conventions](https://docs.netlify.com/configure-builds/manage-dependencies/#npm) (`NPM_TOKEN`) for private modules

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu Were you able to resolve this npm install error/warning? I _believe_ that since netlify takes care of package installation we need to [follow their naming conventions](https://docs.netlify.com/configure-builds/manage-dependencies/#npm) (`NPM_TOKEN`) for private modules

Oh, that could very well be it. Thanks @mattanglin, I'll give that a try.

**Comment by mattanglin** [comment]
@yvonnetangsu Ok. I just realized what the error was you were referring to in the above comment. The Eslint workflow is failing since it can't install the private fa repo. We'll need to update the lint workflow yml config to include the npm token in the job env and add the token itself as a repository secret. Happy to take care of it myself or walk you through it if you'd prefer 👍🏼

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu Ok. I just realized what the error was you were referring to in the above comment. The Eslint workflow is failing since it can't install the private fa repo. We'll need to update the lint workflow yml config to include the npm token in the job env and add the token itself as a repository secret. Happy to take care of it myself or walk you through it if you'd prefer 👍🏼

Ohh - thanks, @mattanglin. I just saw your comment, and tried to change the token name to FONTAWESOME_NPM_AUTH_TOKEN (after trying NPM_TOKEN). Sure, would love to get your help on this. I'll study your changes and ping you with questions 😄

File: .github/workflows/eslint.yml [hunk 1]
```diff
@@ -10,5 +10,11 @@ jobs:
...
       - name: Install modules
         run: npm install
+        env:
+          # The following var needs to correspond with what's used in the .npmrc file for fa pro
+          FONTAWESOME_NPM_AUTH_TOKEN: ${{ secrets.FA_NPM_TOKEN }}
       - name: Run ESLint
         run: npm run lint
+        env:
+          # The following var needs to correspond with what's used in the .npmrc file for fa pro
+          FONTAWESOME_NPM_AUTH_TOKEN: ${{ secrets.FA_NPM_TOKEN }}
```
Comment (mattanglin) [inline .github/workflows/eslint.yml:20]: @yvonnetangsu We needed to include the fa npm token in the lint workflow that runs in github. I added the token as an [encrypted repository secret](https://docs.github.com/en/actions/reference/encrypted-secrets) so we could access here

Comment (yvonnetangsu) [reply .github/workflows/eslint.yml:20]: TIL - thanks, @mattanglin !

**Review by mattanglin — APPROVED** [review]
🎉

**Comment by yvonnetangsu** [comment]
@mattanglin I do notice something strange, https://deploy-preview-148--stanford-alumni.netlify.app/
in the main nav, if you click on "Event" -> "All Events", it goes to a weird URL `https://default/events/` and not resolving it as usual. I noticed that some other internal links on the page works normally e.g., the "Explore your new membership" big red button below the hero.

This seems to only happen in this PR, and not in the itinerary one for example and has never happened before, so I suspect it's related to the FA PRO installation, but I'm not really sure why 🤔  I'll investigate a bit but might need your help on this 🙏🏼

**Comment by yvonnetangsu** [comment]
Looks like the PRO icons are working so that's good, but internal links in nav are still broken....weird 🤔 
https://deploy-preview-148--stanford-alumni.netlify.app/test/yvonne/test-basic-page/

**Comment by yvonnetangsu** [comment]
I inspected the weird link in the google inspector, and it says the `cached_url` that get passed from Storyblok Link component is  `/default/events/`, I looked on Storyblok and it was just `/events/`
Screen Shot 2021-09-02 at 6 43 28 PM

Tagging @sherakama too in case you know something about this weirdness

File: package.json [hunk 2]
```diff
@@ -4,10 +4,15 @@
   "version": "0.1.0",
   "author": "Stanford University | Alumni Association",
   "dependencies": {
+    "@fortawesome/fontawesome-pro": "^5.15.4",
```
Comment (sherakama) [inline package.json:7]: Where did we get the pro license from?

Comment (yvonnetangsu) [reply package.json:7]: Michele gave me the auth token via DM - SAA has a subscription. I added it to Netlify.

Comment (yvonnetangsu) [reply package.json:7]: While working on this, my local repo broke again with that weird maximum all out of range error with Webpack (the one that you mentioned also had, and had to pull down this repo fresh again). I don't think it's related to the work in this branch though.

**Comment by mattanglin** [comment]
> @mattanglin I do notice something strange, https://deploy-preview-148--stanford-alumni.netlify.app/
> in the main nav, if you click on "Event" -> "All Events", it goes to a weird URL `https://default/events/` and not resolving it as usual. I noticed that some other internal links on the page works normally e.g., the "Explore your new membership" big red button below the hero.
> 
> This seems to only happen in this PR, and not in the itinerary one for example and has never happened before, so I suspect it's related to the FA PRO installation, but I'm not really sure why 🤔 I'll investigate a bit but might need your help on this 🙏🏼

That is weird! I can take a look. It surprises me that fontawesome would affect how links are being generated... 🤔

**Comment by yvonnetangsu** [comment]
> > @mattanglin I do notice something strange, https://deploy-preview-148--stanford-alumni.netlify.app/
> > in the main nav, if you click on "Event" -> "All Events", it goes to a weird URL `https://default/events/` and not resolving it as usual. I noticed that some other internal links on the page works normally e.g., the "Explore your new membership" big red button below the hero.
> > This seems to only happen in this PR, and not in the itinerary one for example and has never happened before, so I suspect it's related to the FA PRO installation, but I'm not really sure why 🤔 I'll investigate a bit but might need your help on this 🙏🏼
> 
> That is weird! I can take a look. It surprises me that fontawesome would affect how links are being generated... 🤔

Yeah, doesn't make sense to me - thank you, Matt!

**Comment by mattanglin** [comment]
> > > @mattanglin I do notice something strange, https://deploy-preview-148--stanford-alumni.netlify.app/
> > > in the main nav, if you click on "Event" -> "All Events", it goes to a weird URL `https://default/events/` and not resolving it as usual. I noticed that some other internal links on the page works normally e.g., the "Explore your new membership" big red button below the hero.
> > > This seems to only happen in this PR, and not in the itinerary one for example and has never happened before, so I suspect it's related to the FA PRO installation, but I'm not really sure why 🤔 I'll investigate a bit but might need your help on this 🙏🏼
> > 
> > 
> > That is weird! I can take a look. It surprises me that fontawesome would affect how links are being generated... 🤔
> 
> Yeah, doesn't make sense to me - thank you, Matt!

Well, the reason the link isn't working is likely due to some site or base url issue somewhere as (aside from a long list of classes) I'm seeing the following rendered source for menu links:
image

digging into the component hierarchy now...

**Comment by sherakama** [comment]
One thing that I would ask is that you document the fact that you need the FA token in your environment in order to install and update this package now. Please document it in the readme and socialize it with all the devs/vendors.

**Comment by yvonnetangsu** [comment]
> I inspected the weird link in the google inspector, and it says the `cached_url` that get passed from Storyblok Link component is `/default/events/`, I looked on Storyblok and it was just `/events/`
> Screen Shot 2021-09-02 at 6 43 28 PM

@mattanglin
Yeah, I also saw the default/events from Components inspector on Google, but when I looked at the json data that get passed on the Storyblok side, it says just /events/. So somewhere it was changed 🤔 
@mattanglin

**Comment by yvonnetangsu** [comment]
> One thing that I would ask is that you document the fact that you need the FA token in your environment in order to install and update this package now. Please document it in the readme and socialize it with all the devs/vendors.

Will do, good idea, Shea! I'll add the readme right before it's ready to merge.

**Comment by mattanglin** [comment]
> > I inspected the weird link in the google inspector, and it says the `cached_url` that get passed from Storyblok Link component is `/default/events/`, I looked on Storyblok and it was just `/events/`
> > Screen Shot 2021-09-02 at 6 43 28 PM
> 
> @mattanglin
> Yeah, I also saw the default/events from Components inspector on Google, but when I looked at the json data that get passed on the Storyblok side, it says just /events/. So somewhere it was changed 🤔

Hmmm... Yeah, I'm not accustomed to the response structure of Storyblok url fields quite yet, but the component we're using to render them is what's causing the double leading slash:
https://github.com/SU-SWS/saa_alumni/blob/main/src/utilities/sbLink.js#L54

**Comment by sherakama** [comment]
Dumping the output that goes through SBLink I see a difference for the same ID:

First few times through:
```
{
  id: 'a8112112-016d-496a-8958-6b8122e50493',
  url: '',
  linktype: 'story',
  fieldtype: 'multilink',
  cached_url: '/default/events/',
  prep: true
}
```

Later: 
```
{
  id: 'a8112112-016d-496a-8958-6b8122e50493',
  url: '',
  linktype: 'story',
  fieldtype: 'multilink',
  cached_url: 'events/'
}
```

What is that prep key?

**Comment by yvonnetangsu** [comment]
> Dumping the output that goes through SBLink I see a difference for the same ID:
> 
> First few times through:
> 
> ```
> {
>   id: 'a8112112-016d-496a-8958-6b8122e50493',
>   url: '',
>   linktype: 'story',
>   fieldtype: 'multilink',
>   cached_url: '/default/events/',
>   prep: true
> }
> ```
> 
> Later:
> 
> ```
> {
>   id: 'a8112112-016d-496a-8958-6b8122e50493',
>   url: '',
>   linktype: 'story',
>   fieldtype: 'multilink',
>   cached_url: 'events/'
> }
> ```
> 
> What is that prep key?

Interesting.... I don't recall seeing the prep key before.

**Comment by sherakama** [comment]
I don't see any documentation for it nor do I see anything in our codebase for this.
https://www.storyblok.com/docs/api/content-delivery#core-resources/links/the-link-object

Must be something coming from storyblok.

Could it be something with an outdated schema or something?

**Comment by yvonnetangsu** [comment]
> I don't see any documentation for it nor do I see anything in our codebase for this.
> https://www.storyblok.com/docs/api/content-delivery#core-resources/links/the-link-object
> 
> Must be something coming from storyblok.
> 
> Could it be something with an outdated schema or something?

I checked the SAA main nav last night to see if there are any weird changes, but they all looked fine. The live site seems to be fine as well. First time seeing this 😬

**Comment by mattanglin** [comment]
> > > I inspected the weird link in the google inspector, and it says the `cached_url` that get passed from Storyblok Link component is `/default/events/`, I looked on Storyblok and it was just `/events/`
> > > Screen Shot 2021-09-02 at 6 43 28 PM
> > 
> > 
> > @mattanglin
> > Yeah, I also saw the default/events from Components inspector on Google, but when I looked at the json data that get passed on the Storyblok side, it says just /events/. So somewhere it was changed 🤔
> 
> Hmmm... Yeah, I'm not accustomed to the response structure of Storyblok url fields quite yet, but the component we're using to render them is what's causing the double leading slash:
> https://github.com/SU-SWS/saa_alumni/blob/main/src/utilities/sbLink.js#L54

Ok, digging a little deeper the `/default` portion of the `cached_url` 

> > I don't see any documentation for it nor do I see anything in our codebase for this.
> > https://www.storyblok.com/docs/api/content-delivery#core-resources/links/the-link-object
> > Must be something coming from storyblok.
> > Could it be something with an outdated schema or something?
> 
> I checked the SAA main nav last night to see if there are any weird changes, but they all looked fine. The live site seems to be fine as well. First time seeing this 😬

I couldn't find much useful info on storyblok links in general, specifically usage of the `cached_url` property. How did we come to use that? The leading `/default` portion seems to be i18n related.

**Comment by yvonnetangsu** [comment]
> > > I don't see any documentation for it nor do I see anything in our codebase for this.
> > > https://www.storyblok.com/docs/api/content-delivery#core-resources/links/the-link-object
> > > Must be something coming from storyblok.
> > > Could it be something with an outdated schema or something?
> > 
> > 
> > I checked the SAA main nav last night to see if there are any weird changes, but they all looked fine. The live site seems to be fine as well. First time seeing this 😬
> 
> I couldn't find much useful info on storyblok links in general, specifically usage of the `cached_url` property. How did we come to use that? The leading `/default` portion seems to be i18n related.



> > > I don't see any documentation for it nor do I see anything in our codebase for this.
> > > https://www.storyblok.com/docs/api/content-delivery#core-resources/links/the-link-object
> > > Must be something coming from storyblok.
> > > Could it be something with an outdated schema or something?
> > 
> > 
> > I checked the SAA main nav last night to see if there are any weird changes, but they all looked fine. The live site seems to be fine as well. First time seeing this 😬
> 
> I couldn't find much useful info on storyblok links in general, specifically usage of the `cached_url` property. How did we come to use that? The leading `/default` portion seems to be i18n related.

Interesting about the i18n 🤔 
There's a little bit of info here:
https://www.storyblok.com/faq/link-object-history
We found that when we add the Storyblok Link field, if someone selects the "internal" link option, it will return a cached_url with the linktype "story". If it's external, then it returns a url instead of a `cached_url` (You probably know this part since you've seen our SbLink component, which was created to work with the Storyblok's Link field). So the reason that we use cached_url wasn't really a choice, but it was a result of seeing what data was returned when a user added a certain type of link and having to work with that. Not sure if that's the answer you're looking for though, @mattanglin 😂

**Comment by mattanglin** [comment]
@yvonnetangsu Not sure what to do with this one at this point. The links are breaking and I can't run it locally anymore 🤷🏼

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu Not sure what to do with this one at this point. The links are breaking and I can't run it locally anymore 🤷🏼

We could let Michele know there is some compatibility issue with the PRO icon installation perhaps. We did give them the extra new free icons (probably hundreds more than the original batch), so that could be enough to make our clients happy 🤔

**Comment by mattanglin** [comment]
> > @yvonnetangsu Not sure what to do with this one at this point. The links are breaking and I can't run it locally anymore 🤷🏼
> 
> We could let Michele know there is some compatibility issue with the PRO icon installation perhaps. We did give them the extra new free icons (probably hundreds more than the original batch), so that could be enough to make our clients happy 🤔

That might be a better solution.
Here's another possible thing to try based purely on my own speculation: You _might need to specify a subset of icons from the fontawesome pro catalog and install them individually_. I would guess that the way we're currently implementing fontawesome pro is bringing in the entire library (which is also terrible for our bundle) and that might be what's causing all the memory issues during build.
Still unsure how that would affect the sb links, though 🤷🏼

**Comment by sherakama** [comment]
Yeah, this has a smell to it that doesn't come from fontawesome. 

I suspect something else moved during the install of this package that is causing this issue. I'd like to spend a bit more time troubleshooting this before we point the blame at fontawesome.

**Comment by yvonnetangsu** [comment]
> Yeah, this has a smell to it that doesn't come from fontawesome.
> 
> I suspect something else moved during the install of this package that is causing this issue. I'd like to spend a bit more time troubleshooting this before we point the blame at fontawesome.

Yeah, agree with you. It could be the way I installed it initially causing issues with package-lock.json? It might have moved other things in different packages 😬 .

**Comment by sherakama** [comment]
My update seems to have fixed the default link issue.

Something in one of the other packages must have moved our cheese...

**Comment by yvonnetangsu** [comment]
> My update seems to have fixed the default link issue.
> 
> Something in one of the other packages must have moved our cheese...

Oh that's great. And running npm run develop doesn't break your local, @sherakama? It broke both mine and Matt's. Let me try again.

**Comment by sherakama** [comment]
> Oh that's great. And running npm run develop doesn't break your local, @sherakama? It broke both mine and Matt's. Let me try again.

I did get it to run but had to `gatsby clean` first

File: src/components/cta/ctaLink.js [hunk 3]
```diff
@@ -77,7 +77,8 @@ const CtaLink = React.forwardRef(
                   iconType={type}
                   iconStyle={iconStyle}
                   isOutline={isOutlineFaIcon}
-                  className="su-mr-06em su-backface-hidden su-text-black-80"
+                  className="su-mr-06em su-backface-hidden su-text-black-80 su-flex-shrink-0"
+                  fixedWidth
```
Comment (yvonnetangsu) [inline src/components/cta/ctaLink.js:81]: Added the fixedWidth prop that comes with React FontAwesome, so all icon takes up a fixed width for better vertical alignment (when they're stack on top of each other when used as a leading icon for CTA links)

File: package.json [hunk 4]
```diff
@@ -17,6 +22,7 @@
     "decanter-react": "^1.0.7",
     "gatsby": "^3.4.2",
     "gatsby-link": "^3.12.0",
+    "gatsby-plugin-fontawesome-css": "^1.1.0",
```
Comment (yvonnetangsu) [inline package.json:25]: Added this plugin to prevent icon size blowing up when page first load. It's a common problem for Fontawesome React/Gatsby.
https://github.com/FortAwesome/react-fontawesome/issues/53
https://www.npmjs.com/package/gatsby-plugin-fontawesome-css

**Comment by yvonnetangsu** [comment]
@sherakama Your commit un-broke my repo 🎉  and I added some minor touch ups. Ready for your review - I've updated steps to test.

File: src/components/simple/faIcon.js [hunk 5]
```diff
@@ -16,7 +20,7 @@ const FaIcon = ({
   ...props
 }) => {
   // Add all 3 types of FA icons to the library so you can use any of them
-  library.add(fab, far, fas);
+  library.add(fab, far, fas, fad, fal, pfar, pfas);
```
Comment (sherakama) [inline src/components/simple/faIcon.js:23]: @yvonnetangsu, 

@mattanglin has a good point about adding ALL the library options and our bundle. This does seem like A LOT of icons to add to the site. Perhaps we can come up with a strategy to include just the ones we need? I suspect he is right about it contributing to the memory issues.

Comment (yvonnetangsu) [reply src/components/simple/faIcon.js:23]: @sherakama Hmm... I think they want to be able to use all the available icons, and if we narrow it down, it might be hundreds still 😬  I could check with them to see maybe they could leave out the light ones or the duotone ones, that should help a bit?

Comment (yvonnetangsu) [reply src/components/simple/faIcon.js:23]: I just pinged Michele to see if they're ok with one of these options:

1. Remove all icons of one or more certain styles (eg, remove the "light" ones, or "duotone" ones)

2. Provide us with a full list of icons that you would use, so we'll only import those one by one into the project (eg, if you have 50 or even 100 that you'll end up using)

Comment (yvonnetangsu) [reply src/components/simple/faIcon.js:23]: Ok, got a plan. Michele agree to keep only the solids and lights. And I can just import the few brand ones that they use in the local footer. That should be a pretty big saving.

**Review by sherakama — APPROVED** [review]
Thanks for all the updates. One last request, please document the NPM token in the installation part of the README.md 

@mattanglin any additional feedback?

**Comment by yvonnetangsu** [comment]
> Thanks for all the updates. One last request, please document the NPM token in the installation part of the README.md
> 
> @mattanglin any additional feedback?

Oh good reminder. I'll do that 😄

File: README.md [hunk 6]
```diff
@@ -18,19 +18,23 @@ Installation
...
-* Manually add the access token to .env.production and remove the `GATSBY_HOT_LOADER=fast-refresh line`*
+cp example.env .env.production
+* Get the Storyblok public access token *
+* Get the FontAwesome NPM token from the Netlify site settings in environment variables *
+* Manually add the access tokens to .env.production and remove the `GATSBY_HOT_LOADER=fast-refresh line`*
 nvm use
-npm install
+FONTAWESOME_NPM_AUTH_TOKEN=MYFATOKEN npm install
 gatsby build
 ```
```
Comment (yvonnetangsu) [inline README.md:40]: @sherakama I've updated the readme and the example.env file. Going to merge this in.

File: src/components/simple/faIcon.js (hunk 5 above)
Comment (mattanglin) [reply src/components/simple/faIcon.js]: Sorry I missed all the new feedback here. With this updated we can dive into the build using [bundle analyzer](https://www.gatsbyjs.com/plugins/gatsby-plugin-webpack-bundle-analyser-v2/) to see how the icon sets are affecting bundle size or if it's just the dev memory overhead

Comment (yvonnetangsu) [reply src/components/simple/faIcon.js]: Oh cool. That should be interesting 😄

File: README.md (hunk 6 above)
Comment (sherakama) [reply README.md:40]: ty
